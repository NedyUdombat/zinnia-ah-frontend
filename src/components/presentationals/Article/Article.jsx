import React from 'react';
import PropTypes from 'prop-types';

//style
import 'semantic-ui-css/semantic.min.css';
import './Article.scss';

import Title from '../Title/Title';
import UserThumbnail from '../UserThumbnail/UserThumbnail';

const Article = props => {
  const { user, article } = props;
  return (
    <div className="article">
      <div className="thumbnail">
        <a href={article.slug}>
          <img src={article.thumbnail} />
        </a>
      </div>
      <div className="description">
        <a href={article.slug}>
          <Title content={article.title} className="ui header" />
        </a>
        <p>{article.description}</p>
        <div className="footer">
          <UserThumbnail
            name={`${user.firstName} ${user.lastName}`}
            info={`@${user.username}`}
            image={user.image}
            url={user.username}
          />

          <a href={article.slug} className="btn btn-white read-more">
            Read More
          </a>
        </div>
      </div>
    </div>
  );
};

Article.propTypes = {
  user: PropTypes.any.isRequired,
  article: PropTypes.any.isRequired,
};

export default Article;
