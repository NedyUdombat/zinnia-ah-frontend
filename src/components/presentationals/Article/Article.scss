.articles {
  display: flex;
  flex-wrap: wrap;
  margin: 2rem auto;
  width: 95%;

  .article {
    display: flex;
    margin: 1rem auto;
    min-width: 35rem;
    width: 50%;
    min-height: 13rem;
    transition: transform 13s ease-in-out;
  }

  .article .thumbnail {
    flex-basis: 35%;
    overflow: hidden;

    img {
      height: 100%;
      width: 100%;
      transition: transform 3s ease-in-out;

      &:hover {
        transform: scale(1.5);
        transition: transform 3s ease;
        filter: drop-shadow(16px 16px 20px red) sepia(90%);
      }
    }
  }

  .article .description {
    flex-basis: 60%;
    padding: 10px;

    .header {
      color: #000;
      font-family: Lora;
      font-size: 2rem;
      font-style: normal;
      font-weight: bold;
      line-height: 31px;
      width: 300px;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      padding-bottom: 3px;
    }

    p {
      color: #333;
      font-family: Lato;
      font-size: 1.05rem;
      font-style: normal;
      font-weight: normal;
      line-height: 17px;
      height: 5rem;
      display: -webkit-box;
      -webkit-line-clamp: 4;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    .footer {
      align-items: stretch;
      display: flex;
      flex-direction: row;
      padding-top: 0px;

      .read-more {
        flex-basis: 30%;
        padding-top: 10px;
        width: 82px;
      }
    }
  }
}
