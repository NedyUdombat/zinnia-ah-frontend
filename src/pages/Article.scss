.article-container {
  display: flex;
  justify-content: space-around;
  padding: 0% 6%;

  .sidebar {
    display: flex;
    flex-direction: column;
    height: 6.25rem;
    justify-content: space-between;
    width: 25%;
  }

  .links {
    width: max-content;
  }

  .content {
    text-align: justify;
  }

  .author {
    display: flex;
    width: 25%;
  }
}
