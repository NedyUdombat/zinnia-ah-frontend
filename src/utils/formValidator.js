/**
 * validate inputs
 * usage: const errors = validate({ email: 'user@mail.com', password: 'aNewSecret' }, exitsingErrors);
 */
export const validate = (values, validationErrors) => {
  const newErrors = [];
  const oldErrors = [...validationErrors];
  const keys = Object.keys(values);

  const username = value => {
    if (isFieldEmpty('username', value)) {
      return;
    }

    if (trim(value).length < 3) {
      newErrors.push(lengthError('username', 3));
    } else {
      const index = oldErrors.indexOf(lengthError('username', 3));

      oldErrors.includes(lengthError('username', 3))
        ? oldErrors.splice(index, 1)
        : '';
    }

    if (!trim(value).match(/^([a-zA-Z0-9]+)$/)) {
      newErrors.push(alphanumError('username'));
    } else {
      const index = oldErrors.indexOf(alphanumError('username'));

      oldErrors.includes(alphanumError('username'))
        ? oldErrors.splice(index, 1)
        : '';
    }
  };

  const password = value => {
    if (isFieldEmpty('password', value)) {
      return;
    }

    if (trim(value).length < 8) {
      newErrors.push(lengthError('password', 8));
    } else {
      const index = oldErrors.indexOf(lengthError('password', 8));

      oldErrors.includes(lengthError('password', 8))
        ? oldErrors.splice(index, 1)
        : '';
    }

    if (!trim(value).match(/^([a-zA-Z0-9]+)$/)) {
      newErrors.push(alphanumError('password'));
    } else {
      const index = oldErrors.indexOf(alphanumError('password'));

      oldErrors.includes(alphanumError('password'))
        ? oldErrors.splice(index, 1)
        : '';
    }
  };

  const email = value => {
    if (isFieldEmpty('email', value)) {
      return;
    }

    const pattern = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
    const isEmail = trim(value).match(pattern);
    if (!isEmail) {
      newErrors.push(emailError);
    } else {
      const index = oldErrors.indexOf(emailError);

      oldErrors.includes(emailError) ? oldErrors.splice(index, 1) : '';
    }
  };

  const isFieldEmpty = (field, value) => {
    if (trim(value).length < 1) {
      newErrors.push(isRequiredError(field));
      return true;
    }
    const index = oldErrors.indexOf(isRequiredError(field));

    oldErrors.includes(isRequiredError(field))
      ? oldErrors.splice(index, 1)
      : '';

    return false;
  };

  keys.forEach(key => {
    switch (key) {
      case 'username':
        username(values[key]);
        break;

      case 'email':
        email(values[key]);
        break;

      case 'password':
        password(values[key]);
        break;

      default:
        break;
    }
  });

  return [...new Set(oldErrors.concat(newErrors))];
};

export const emailError = 'email must be a valid email';

export const alphanumError = field => {
  return `${field} can only contain alphanumerics`;
};

export const isRequiredError = field => {
  return `${field} is required`;
};

export const lengthError = (field, requiredLength) => {
  return `${field} must be at least ${requiredLength} characters long`;
};

export const trim = word => word.replace(/([ ])+/g, '');
