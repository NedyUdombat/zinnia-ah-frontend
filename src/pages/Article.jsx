import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Icon, Item } from 'semantic-ui-react';
import moment from 'moment';

// components
import Navbar from '../components/presentationals/Navbar/Navbar';

//modules
import { getSingleArticle } from '../store/modules/article';
import Loader from '../components/presentationals/Loader/Loader';
import Title from '../components/presentationals/Title/Title';
import Author from '../components/presentationals/Author/Author';

//stylesheets
import './Article.scss';

export class Article extends Component {
  componentDidMount() {
    this.props.getSingleArticle(this.props.match.params.articleId);
  }
  render() {
    const { article } = this.props;
    return (
      <Fragment>
        {this.props.isLoading && (
          <Loader size="large" text="loading, please wait" />
        )}
        <Navbar profileUrl={this.props.profileUrl} />
        <div className="d-flex article-container" data-test="article">
          <div className="author">
            {article.author && <Author profile={article.author} />}
          </div>
          <div className="content">
            <Item>
              <Title content={article.title} className="title-lg" />
              {moment(article.createdAt).fromNow()}
              <Item.Content>
                <Item.Image src={article.imageThumbnail} size="massive" />
                <br />
                <Item.Description>{article.body}</Item.Description>
              </Item.Content>
            </Item>
          </div>
          <div className="sidebar">
            <a
              className="links"
              target="_blank"
              rel="noopener noreferrer"
              href={`https://twitter.com/intent/tweet?text=${
                article.title
              }&url=${location.href}`}
            >
              <Icon name="twitter" size="large" />
            </a>
            <a>
              <Icon name="facebook" size="large" />
            </a>
            <a
              className="links"
              rel="noopener noreferrer"
              target="_blank"
              href={`mailto:''?subject=${article.title}&body=${location.href}`}
            >
              <Icon name="mail" size="large" />
            </a>
          </div>
        </div>
      </Fragment>
    );
  }
}

Article.propTypes = {
  article: PropTypes.object.isRequired,
  getSingleArticle: PropTypes.func,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      articleId: PropTypes.string.isRequired,
    }),
  }),
  profileUrl: PropTypes.string,
};

Article.defaultProps = {
  profileUrl: '',
};

const mapStateToProps = state => ({
  isLoading: state.article.isLoading,
  article: state.article.article,
});

export default connect(
  mapStateToProps,
  { getSingleArticle },
)(Article);
