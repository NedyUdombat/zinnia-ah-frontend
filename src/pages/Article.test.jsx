import React from 'react';
import { shallow } from 'enzyme';
import checkPropTypes from 'check-prop-types';

import { Article } from './Article';
import { findByTestAttribute } from '../utils/testHelpers';

const props = {
  match: {
    params: {
      articleId: '',
    },
  },
  getSingleArticle: jest.fn(),
  isLoading: false,
  article: {},
};

describe('Article', () => {
  it('should render without errors', () => {
    const wrapper = shallow(<Article {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(findByTestAttribute(wrapper, 'article').length).toEqual(1);
  });

  it('should receive props', () => {
    const expectedProps = { ...props };

    const propsError = checkPropTypes(
      Article.propTypes,
      expectedProps,
      'props',
      Article,
    );

    expect(propsError).toBeUndefined();
  });
});
