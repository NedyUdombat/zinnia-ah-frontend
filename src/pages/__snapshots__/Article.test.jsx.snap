// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should render without errors 1`] = `
<Fragment>
  <Navbar
    profileUrl=""
  />
  <div
    className="d-flex article-container"
    data-test="article"
  >
    <div
      className="author"
    />
    <div
      className="content"
    >
      <Item>
        <Title
          className="title-lg"
        />
        a few seconds ago
        <ItemContent>
          <ItemImage
            size="massive"
          />
          <br />
          <ItemDescription />
        </ItemContent>
      </Item>
    </div>
    <div
      className="sidebar"
    >
      <a
        className="links"
        href="https://twitter.com/intent/tweet?text=undefined&url=http://localhost/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <Icon
          as="i"
          name="twitter"
          size="large"
        />
      </a>
      <a>
        <Icon
          as="i"
          name="facebook"
          size="large"
        />
      </a>
      <a
        className="links"
        href="mailto:''?subject=undefined&body=http://localhost/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <Icon
          as="i"
          name="mail"
          size="large"
        />
      </a>
    </div>
  </div>
</Fragment>
`;
